In [62]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [63]:
df = pd.read_csv('/content/drive/MyDrive/Machine Learning/Projects/Real Estate Capstone Project/DataSets/Feature Engineering and Selection/before_feature_real_estate.csv')
df_copy = df.copy()
df.head()

,bhk,price,which_floor,bathrooms,balcony,parking,buildup_area_sqft,is_ready_to_move,furnishing,age_category,location_zone
0,3.0,2.05,Middle,3.0,3.0,1.0,1862.0,1,Semi Furnished,Mid Age,Western Suburbs East
1,1.0,1.33,Lower,2.0,0.0,1.0,790.0,1,Unfurnished,New,Western Suburbs East
2,3.0,7.30,Lower,2.0,0.0,2.0,1430.0,1,Unfurnished,Mid Age,Western Suburbs East
3,3.0,7.60,Higher,4.0,0.0,3.0,2195.0,1,Semi Furnished,Relatively New,Western Suburbs East
4,2.0,3.20,Lower,2.0,1.0,1.0,960.0,1,Unfurnished,Relatively New,South Mumbai


In [64]:
df2 = pd.read_csv('/content/drive/MyDrive/Machine Learning/Projects/Real Estate Capstone Project/DataSets/Outliers Removed/Outliers_removed_real_estate.csv')
df2.head()

,bhk,price,age_of_property,which_floor,project_name,bedrooms,bathrooms,balcony,parking,source,...,buildup_area_sqft,carpet_area_sqft,is_ready_to_move,furnishing,price_per_sqft,property_score,locationadvantages,topfacilities,age_category,log_price
0,3.0,2.05,15.0,Middle,Moroccan Cooperative,3.0,3.0,3.0,1.0,kaggle,...,1862.0,1425.0,True,Semi Furnished,11009.667025,4,NaN,NaN,Mid Age,1.115142
1,1.0,1.33,1.0,Lower,Rajesh White City,1.0,2.0,0.0,1.0,kaggle,...,790.0,479.0,True,Unfurnished,16835.443038,1,NaN,NaN,New,0.845868
2,3.0,7.30,10.0,Lower,Oberoi Esquire,3.0,2.0,0.0,2.0,kaggle,...,1430.0,1330.0,True,Unfurnished,51048.951049,3,NaN,NaN,Mid Age,2.116256
3,3.0,7.60,4.0,Higher,Oberoi Esquire,3.0,4.0,0.0,3.0,kaggle,...,2195.0,1400.0,True,Semi Furnished,34624.145786,5,NaN,NaN,Relatively New,2.151762
4,2.0,3.20,3.0,Lower,Ajmera Zeon,2.0,2.0,1.0,1.0,kaggle,...,960.0,860.0,True,Unfurnished,33333.333333,3,NaN,NaN,Relatively New,1.435085


## Inference using location_zone

In [65]:
df['which_floor'] = df['which_floor'].replace({'Lower':'0', 'Middle':1, 'Higher':2, 'Unknown': 0})
df['furnishing'] = df['furnishing'].replace({'Semi Furnished':2, 'Unfurnished':1, 'Fully Furnished':3, 'Unknown':0})

df['which_floor'] = pd.to_numeric(df['which_floor'])

/tmp/ipykernel_3673/3089329485.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['furnishing'] = df['furnishing'].replace({'Semi Furnished':2, 'Unfurnished':1, 'Fully Furnished':3, 'Unknown':0})


In [66]:
X_df = pd.get_dummies(df, columns=['age_category', 'location_zone'], drop_first=True)

In [67]:
X_df

,bhk,price,which_floor,bathrooms,balcony,parking,buildup_area_sqft,is_ready_to_move,furnishing,age_category_New,age_category_Old,age_category_Relatively New,location_zone_Outskirts,location_zone_South Mumbai,location_zone_Thane Navi Mumbai,location_zone_Western Suburbs,location_zone_Western Suburbs East
0,3.0,2.05,1,3.0,3.0,1.0,1862.0,1,2,False,False,False,False,False,False,False,True
1,1.0,1.33,0,2.0,0.0,1.0,790.0,1,1,True,False,False,False,False,False,False,True
2,3.0,7.30,0,2.0,0.0,2.0,1430.0,1,1,False,False,False,False,False,False,False,True
3,3.0,7.60,2,4.0,0.0,3.0,2195.0,1,2,False,False,True,False,False,False,False,True
4,2.0,3.20,0,2.0,1.0,1.0,960.0,1,1,False,False,True,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10031,1.0,1.17,1,1.0,0.0,1.0,400.0,1,1,False,False,True,False,True,False,False,False
10032,2.0,2.00,0,2.0,0.0,1.0,753.0,0,0,False,False,True,False,False,False,False,False
10033,2.0,1.84,1,2.0,0.0,2.0,851.0,1,2,False,False,True,False,False,False,False,False
10034,2.0,0.95,0,2.0,0.0,0.0,974.0,0,0,False,False,True,False,False,True,False,False


In [68]:
X = X_df.drop(columns=['price'])
y = X_df['price']

In [69]:
y_log = np.log1p(y)

In [70]:
from sklearn.preprocessing import StandardScaler

# Re-create X after fixing the dataframe
X_df = pd.get_dummies(df, columns=['age_category', 'location_zone'], drop_first=True)
X = X_df.drop(columns=['price'])

scaler = StandardScaler()
X_scaled_array = scaler.fit_transform(X)
X_scaled = pd.DataFrame(X_scaled_array, columns=X.columns)
display(X_scaled.head())

,bhk,which_floor,bathrooms,balcony,parking,buildup_area_sqft,is_ready_to_move,furnishing,age_category_New,age_category_Old,age_category_Relatively New,location_zone_Outskirts,location_zone_South Mumbai,location_zone_Thane Navi Mumbai,location_zone_Western Suburbs,location_zone_Western Suburbs East
0,0.803053,0.149274,0.851945,2.178470,-0.407072,1.085824,0.295706,0.780010,-0.414422,-0.302428,-0.927492,-0.310791,-0.357800,-0.330818,-0.593308,1.963503
1,-1.163023,-1.113892,-0.339709,-0.590767,-0.407072,-0.481485,0.295706,-0.246637,2.413000,-0.302428,-0.927492,-0.310791,-0.357800,-0.330818,-0.593308,1.963503
2,0.803053,-1.113892,-0.339709,-0.590767,0.883318,0.454222,0.295706,-0.246637,-0.414422,-0.302428,-0.927492,-0.310791,-0.357800,-0.330818,-0.593308,1.963503
3,0.803053,1.412440,2.043600,-0.590767,2.173708,1.572683,0.295706,0.780010,-0.414422,-0.302428,1.078177,-0.310791,-0.357800,-0.330818,-0.593308,1.963503
4,-0.179985,-1.113892,-0.339709,0.332312,-0.407072,-0.232938,0.295706,-0.246637,-0.414422,-0.302428,1.078177,-0.310791,2.794859,-0.330818,-0.593308,-0.509294


In [71]:
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LinearRegression, Ridge

kfold = KFold(n_splits=10, shuffle=True, random_state=42)
scores = cross_val_score(LinearRegression(), X_scaled, y_log, cv=kfold, scoring='r2')

In [72]:
scores.mean(), scores.std()

(np.float64(0.8615946893386066), np.float64(0.010300278457438578))

In [73]:
lr = LinearRegression()
ridge = Ridge(alpha=0.0001)

In [74]:
lr.fit(X_scaled,y_log)

LinearRegression()

In [75]:
ridge.fit(X_scaled,y_log)

Ridge(alpha=0.0001)

In [76]:
ridge.coef_.shape

(16,)

In [77]:
coef_df = pd.DataFrame(ridge.coef_.reshape(1,16),columns=X.columns).stack().reset_index().drop(columns=['level_0']).rename(columns={'level_1':'feature',0:'coef'})
coef_df

,feature,coef
0,bhk,0.233047
1,which_floor,-0.019025
2,bathrooms,-0.013006
3,balcony,-0.021139
4,parking,0.032446
5,buildup_area_sqft,0.245755
6,is_ready_to_move,0.022145
7,furnishing,-0.004561
8,age_category_New,0.012972
9,age_category_Old,0.013657


In [78]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

Z = X_df[['buildup_area_sqft','bhk','bathrooms', 'balcony', 'parking', 'furnishing']]

vif = pd.DataFrame({
    'feature': Z.columns,
    'VIF': [variance_inflation_factor(Z.values, i)
            for i in range(Z.shape[1])]
})

print(vif)

             feature        VIF
0  buildup_area_sqft  11.731725
1                bhk  21.607304
2          bathrooms  12.349922
3            balcony   1.400523
4            parking   5.661634
5         furnishing   2.245587


In [79]:
df[['buildup_area_sqft','bhk','bathrooms']].corr()

,buildup_area_sqft,bhk,bathrooms
buildup_area_sqft,1.000000,0.819912,0.642653
bhk,0.819912,1.000000,0.718223
bathrooms,0.642653,0.718223,1.000000


In [80]:
models = {
    'Model_A': X_scaled.columns,

    'Model_B': X_scaled.drop(columns='bathrooms').columns,

    'Model_C': X_scaled.drop(columns='bhk').columns
}

In [85]:
results = []

for name, features in models.items():

    X_subset = X_scaled[features]

    scores = cross_val_score(
        LinearRegression(),
        X_subset,
        y_log,
        cv=5,
        scoring='r2'
    )

    results.append({
        'Model': name,
        'Features': len(features),
        'Mean CV R²': scores.mean(),
        'Std CV R²': scores.std()
    })

In [86]:
results_df = pd.DataFrame(results)

print(results_df.sort_values('Mean CV R²', ascending=False))

     Model  Features  Mean CV R²  Std CV R²
0  Model_A        16    0.861956   0.005766
1  Model_B        15    0.861811   0.005770
2  Model_C        15    0.830026   0.007750


0.861956 - 0.830026 = ~ 0.032
- Removing BHK hurts performance.
- BHK contains useful information that bathrooms cannot fully replace.



In [88]:
X_scaled_v2 = X_scaled.drop(columns='bathrooms')

In [89]:
lr.fit(X_scaled_v2,y_log)

LinearRegression()

In [93]:
coef_df_v2 = pd.DataFrame(lr.coef_.reshape(1,15),columns=X_scaled_v2.columns).stack().reset_index().drop(columns=['level_0']).rename(columns={'level_1':'feature',0:'coef'})
coef_df_v2

,feature,coef
0,bhk,0.225493
1,which_floor,-0.018867
2,balcony,-0.021391
3,parking,0.032413
4,buildup_area_sqft,0.243608
5,is_ready_to_move,0.022087
6,furnishing,-0.003609
7,age_category_New,0.012641
8,age_category_Old,0.013672
9,age_category_Relatively New,0.029498


In [100]:
feature_stds = X[X_scaled_v2.columns].std()

unscaled_coefficients = (
    coef_df_v2.set_index('feature')['coef']
    / feature_stds
)
display(unscaled_coefficients)

,0
feature,
bhk,0.221658
which_floor,-0.023831
balcony,-0.019745
parking,0.041824
buildup_area_sqft,0.000356
is_ready_to_move,0.081221
furnishing,-0.003705
age_category_New,0.035741
age_category_Old,0.049339


In [103]:
feature_stds = X[X_scaled_v2.columns].std()

beta = (
    coef_df_v2
    .set_index('feature')['coef']
    / feature_stds
)

results = []

for feature, b in beta.items():

    # Area: interpret as +100 sqft
    if feature == 'buildup_area_sqft':

        effect_pct = (np.exp(b * 100) - 1) * 100

        interpretation = "+100 sqft"

    # Dummy variables: interpret as 0 -> 1
    elif (
        feature.startswith('location_zone_')
        or feature.startswith('age_category_')
        or feature == 'is_ready_to_move'
    ):

        effect_pct = (np.exp(b) - 1) * 100

        interpretation = "0 -> 1"

    else:

        effect_pct = (np.exp(b) - 1) * 100

        interpretation = "+1 unit"

    results.append({
        'feature': feature,
        'beta': round(b, 6),
        'interpretation': interpretation,
        'price_effect_%': round(effect_pct, 2)
    })

inference_df = pd.DataFrame(results)

display(
    inference_df.sort_values(
        'price_effect_%',
        ascending=False
    )
)

,feature,beta,interpretation,price_effect_%
11,location_zone_South Mumbai,0.445689,0 -> 1,56.16
0,bhk,0.221658,+1 unit,24.81
13,location_zone_Western Suburbs,0.157037,0 -> 1,17.00
5,is_ready_to_move,0.081221,0 -> 1,8.46
9,age_category_Relatively New,0.059160,0 -> 1,6.09
8,age_category_Old,0.049339,0 -> 1,5.06
3,parking,0.041824,+1 unit,4.27
7,age_category_New,0.035741,0 -> 1,3.64
4,buildup_area_sqft,0.000356,+100 sqft,3.63
14,location_zone_Western Suburbs East,0.030029,0 -> 1,3.05


In [105]:
import statsmodels.api as sm

X_inf = X_scaled_v2
X_inf = sm.add_constant(X_inf)

model = sm.OLS(y_log, X_inf).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.862
Method:                 Least Squares   F-statistic:                     4187.
Date:                Mon, 15 Jun 2026   Prob (F-statistic):               0.00
Time:                        12:22:42   Log-Likelihood:                 161.80
No. Observations:               10036   AIC:                            -291.6
Df Residuals:                   10020   BIC:                            -176.2
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
cons

In [106]:
model.pvalues.sort_values()

,0
const,0.000000e+00
bhk,0.000000e+00
buildup_area_sqft,0.000000e+00
location_zone_Outskirts,0.000000e+00
location_zone_South Mumbai,0.000000e+00
location_zone_Thane Navi Mumbai,3.131257e-140
location_zone_Western Suburbs,7.261413e-111
parking,1.521191e-27
age_category_Relatively New,1.476432e-23
is_ready_to_move,2.141961e-17
